python/cuda_parallel/tests/test_reduce.py

```diff
@@ -4,8 +4,13 @@
 
 import numpy
 import pytest
-from numba import cuda
+import random
+import re
+import numba.cuda
+import numba.types
 import cuda.parallel.experimental as cudax
+from cuda.parallel.experimental import _iterators
+from cuda.parallel.experimental import iterators
```

Comment:
question: do we want the submodule to be called `cudax.iterators` or `cudax.itertools`? Initial idea was that `cudax.itertools` gives us a simple transition rule "if you need `itertools.X`, add `cudax.` before that and you get same semantics". One concern about this approach was that it might be confusing, since implementation of, say, `cudax.itertools.repeat` is very different from `itertools.repeat` and not usable outside of cuda.parallel. @rwgk, @shwina, @eriknw what's your preference?

Reply:
It's very early days and we shouldn't consider any decisions we make right now as being a "final" public API.  That being said:

1. I don't think we should worry about being any kind of drop-in replacement for `itertools`. I would avoid `cudax.itertools`.
2. I'm fine with putting iterators in the `cudax.iterators` module.

I don't _particularly_ like the names `repeat`, `count`, and `map`, though. Neither do those functions do the equivalent of their `itertools` counterparts (return true iterators), nor do they perform their named action (repeat doesn't repeat anything, count doesn't count anything). Instead, they are just wrappers around the underlying iterator constructors.

 I would much prefer the more explicit names:

* `repeat` -> `make_constant_iterator`
* `count` -> `make_counting_iterator`
* `map` -> `make_transform_iterator`

At a later time, we can always consider building a higher-level API on top of these.

Reply:
I think @shwina's naming suggestions are great. It'd take me only a few minutes to make those changes. @gevtushenko OK to go ahead?

Reply:
@shwina's suggestions also sound good to me. I might go one step further and wonder why e.g. `make_constant_iterator` would be better than using `ConstantIterator` directly.

Reply:
Looks good to me - thanks!

Reply:
I’m not a Python developer by any means, so still trying to figure out the risks of having `cudax.itertools.count` that must be obvious to you. Please bear with me.

>1. I don't think we should worry about being any kind of drop-in replacement for itertools. I would avoid cudax.itertools.

My hope is that using `cudax.itertools` would help developers leverage existing intuition. Say, in C++, if you know `std::atomic_ref`, it's easy for you to transition to `cuda::atomic_ref` since it has only minor API differences. If you know `std::reduce`, it's easy to transition to `thrust::reduce`. In this examples there are CUDA-specific implementation nuances. Atomic ref has a memory scope, while reduce is a parallel algorithms. Most importantly, semantics of these facilities is similar. The fact that there are differences is reflected in the namespace differences. If we have to tell someone about our iterators, it's easy to do so if they already know itertools repeat, count, and map. If we go with different names, how would we explain them in familiar terms? Would we say `iterators.CountingIterator` is like `itertools.count`? 

> I don't _particularly_ like the names repeat, count, and map, though. Neither do those functions do the equivalent of their itertools counterparts (return true iterators), 

`cudax.iterators.CountingIterator` still isn't true Python iterator. I don't see how calling the module `iterators` instead of `itertools` addresses this concern. Given that we are not allowing usage of these iterators with any algorithms outside of `cuda.parallel`, the fact that these types are not true iterators should remain an implementation detail to users. 

> nor do they perform their named action (repeat doesn't repeat anything, count doesn't count anything). Instead, they are just wrappers around the underlying iterator constructors.

Doesn't `itertools.count` also not counting anything but rather constructing an iterator?

> At a later time, we can always consider building a higher-level API on top of these, including _possibly_ a itertools replacement. 

Would such an `itertools` replacement look something like?

```python
def count(x):
  return CountingIterator(x)
```

Reply:
Thanks for bringing up some really great points.

>  if you know std::atomic_ref, it's easy for you to transition to cuda::atomic_ref since it has only minor API differences. If you know std::reduce, it's easy to transition to thrust::reduce .... Most importantly, semantics of these facilities is similar. 

My concern here is that the semantics of our iterators are not very similar to their `itertools` namesakes - they are fundamentally different things.

While yes, in limited contexts, `cuda.parallel` can look _sort of_ like `itertools`:

```python
>>> functools.reduce(lambda x, y: x + y, itertools.takewhile(lambda x: x < 5, itertools.count(1)))
10
```

In the broader scheme of things, iterators are used in ways we simply cannot (and perhaps _should not_) support today:

```python
>>> import numpy as np
>>> np.fromiter(itertools.cycle([1, 2, 3]), float, 10)
array([1., 2., 3., 1., 2., 3., 1., 2., 3., 1.])
```

In my experience with libraries like CuPy, cuDF - people expect a very similar experience to NumPy, Pandas. Even very minor departures in behaviour and API can be frustrating for the user. Many people will give up after noticing/experiencing a handful of such differences.  By calling ourselves `cudax.itertools`, we are inviting the same expectation. 

> If we have to tell someone about our iterators, it's easy to do so if they already know itertools repeat, count, and map. If we go with different names, how would we explain them in familiar terms? Would we say iterators.CountingIterator is like itertools.count?

We could say that. And by not calling it something like `cudax.itertools.count`, we dramatically reduce the expectation that it is identical/usable in the same contexts.

> cudax.iterators.CountingIterator still isn't true Python iterator. I don't see how calling the module iterators instead of itertools addresses this concern.

This is true. We could consider calling it something like `cudax.lazy.Counter` instead of `cudax.iterators.CountIterator`. But perhaps I should take a big step back here :

## Some higher level thoughts on `cuda.parallel`

At the moment, I think of `cuda.experimental.parallel` as an experiment/playground to see how we can expose the host-side functionality of thrust/CUB to Python. We (or at least I) are learning what kinds of use cases this can enable that existing GPU libraries in Python cannot. 

So how do we make decisions about its design/API at this stage? Personally, I think we should attempt to mimic the underlying thrust/CUB C++ APIs as closely as possible. While that may not be "Pythonic", it does give us _something_ to be consistent with and punt on the high-level API question (a hard one). It also minimizes the amount of wrapper logic we need to add to make things more Pythonic. This is akin to the approach taken by projects like [pylibcudf](https://docs.rapids.ai/api/cudf/latest/developer_guide/pylibcudf/) and [CUDA Python](https://github.com/NVIDIA/cuda-python).

Reply:
I will add that I don't think we should simply _ignore_ the question of what a good high-level API should look like. I have been talking to a number of people to gather perspectives about this, and will continue to do so.

I also think there will be a good deal of community effort involved here: for instance, if we had a device equivalent of the Python [iterator protocol](https://docs.python.org/3/glossary.html#term-iterator), then code like the below actually _could_ work and be extremely performant:

```python
# just an illustration of a _possible_ future
import cupy as np
cp.fromiter(cuda.itertools.cycle([1, 2, 3]), float, 10)
```

If folks are broadly aligned on the idea that we should build a low-level API now and a high-level API on top of that, I can open an issue to collect more thoughts/perspectives.

Reply:
I see your points now! Agreed, let's go with C++ vocabulary for now. I think this discussion also supports selecting `value_type` in favor of `dtype`.